source/docs/software/wpilib-tools/robot-simulation/device-sim.rst

```diff
@@ -0,0 +1,117 @@
+        if (value.getType() == HALValue.kInt) {
+          System.out.println("Value of " + name + " is " + value.getInt());
+        }
+      }
+      CallbackStore store = simEncoder.registerCountCallback(callback);
+
+      store.close(); // cancel the callback
+
+   .. code-tab:: cpp
+
+      HAL_NotifyCallback callback = [](const char* name, void* param, const HALValue* value) {
+        if (value->type == HAL_INT) {
+          wpi::outs() << "Value of " << name << " is " << value->data.v_int << '\n';
+        }
+      };
+      frc::sim::CallbackStore store = simEncoder.RegisterCountCallback(callback);
+      // the callback will be canceled when ``store`` goes out of scope
+
+Simulating Other Devices - The SimDeviceSim Class
+-------------------------------------------------
+
+.. note:: Vendors might implement their connection to the SimDevice API slightly different than described here. They might also provide a simulation class specific for their device class. See your vendor's documentation for more information as to what they support and how.
+
+The ``SimDeviceSim`` (**not ``SimDevice``!**) class is a general device simulation object for devices that aren't core WPILib devices and therefore don't have specific simulation classes - such as vendor devices. These devices will show up in the :guilabel:`Other Devices` tab of the :ref:`SimGUI<docs/software/wpilib-tools/robot-simulation/simulation-gui:Modifying ADXRS450 Inputs>`.
+
+The ``SimDeviceSim`` object is created using a string key identical to the key the vendor used to construct the underlying ``SimDevice`` in their device class. This key is the one that the device shows up with in the :guilabel:`Other Devices` tab, and is typically of the form ``Prefix:Device Name[index]``. If the key contains ports/index/channel numbers, they can be passed as separate arguments to the ``SimDeviceSim`` constructor. The key contains a prefix that is hidden by default in the SimGUI, it can be shown by selecting the :guilabel:`Show prefix` option. Not including this prefix in the key passed to ``SimDeviceSim`` will not match the device!
+
+.. tabs::
+   .. code-tab:: java
+
+      SimDeviceSim device = new SimDeviceSim(deviceKey, index);
+
+   .. code-tab:: cpp
+
+      frc::sim::SimDeviceSim device{deviceKey, index};
+
+Once we have the ``SimDeviceSim``, we can get ``SimValue`` objects representing the device's fields. Type-specific ``SimDouble``, ``SimInt``, ``SimLong``, ``SimBoolean``, and ``SimEnum`` subclasses also exist, and should be used instead of the type-unsafe ``SimValue`` class. These are constructed from the ``SimDeviceSim`` using a string key identical to the one the vendor used to define the field. This key is the one the field appears as in the SimGUI. Attempting to retrieve a ``SimValue`` object outside of simulation or when either the device or field keys are unmatched will return ``null`` - this can cause ``NullPointerException``s in Java or undefined behavior in C++.
```

Comment:
```suggestion
Once we have the ``SimDeviceSim``, we can get ``SimValue`` objects representing the device's fields. Type-specific ``SimDouble``, ``SimInt``, ``SimLong``, ``SimBoolean``, and ``SimEnum`` subclasses also exist, and should be used instead of the type-unsafe ``SimValue`` class. These are constructed from the ``SimDeviceSim`` using a string key identical to the one the vendor used to define the field. This key is the one the field appears as in the SimGUI. Attempting to retrieve a ``SimValue`` object outside of simulation or when either the device or field keys are unmatched will return ``null`` - this can cause ``NullPointerException`` in Java or undefined behavior in C++.
```